[Screen Time Refactoring] Add WKWebView property to detect Screen Time blocking

jesxilin opened the pull request:
#### f88f9f5479fc740fee30965236a5a89ae213598c
<pre>
[Screen Time Refactoring] Add WKWebView property to detect Screen Time blocking
<a href="https://bugs.webkit.org/show_bug.cgi?id=284703">https://bugs.webkit.org/show_bug.cgi?id=284703</a>
<a href="https://rdar.apple.com/140438960">rdar://140438960</a>

Reviewed by Aditya Keerthi.

Created new API and property `_isBlockedByScreenTime` to indicate
if ScreenTime blocking is active.
Added API tests for new property.
Added missing header files for tests `SystemPreview.mm`.

* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _installScreenTimeWebpageController]):
(-[WKWebView observeValueForKeyPath:ofObject:change:context:]):
(-[WKWebView _isBlockedByScreenTime]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
* Tools/TestWebKitAPI/SourcesCocoa.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ScreenTime.mm: Added.
(webViewForScreenTimeTests):
(-[BlockedStateObserver initWithWebView:]):
(-[BlockedStateObserver observeValueForKeyPath:ofObject:change:context:]):
(TEST(ScreenTime, IsBlockedByScreenTimeTrue)):
(TEST(ScreenTime, IsBlockedByScreenTimeFalse)):
(TEST(ScreenTime, IsBlockedByScreenTimeMultiple)):
(TEST(ScreenTime, IsBlockedByScreenTimeKVO)):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SystemPreview.mm:

Canonical link: <a href="https://commits.webkit.org/288240@main">https://commits.webkit.org/288240@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1e1545fe58a258b17868783f1d717b6c81f9ae0c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/82089 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/1619 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/36052 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/86649 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/33127 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/84195 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/1656 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/9446 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/63994 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/21717 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/85159 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/1241 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/74692 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/44275 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/1143 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/28871 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/31546 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/72443 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/29483 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/88085 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/9333 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/6669 "Found 2 new test failures: http/wpt/mediarecorder/record-96KHz-sources.html imported/w3c/web-platform-tests/mediacapture-record/MediaRecorder-no-sink.https.html (failure)") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/72366 "Found 4 new test failures: compositing/animation/repaint-after-clearing-shared-backing.html imported/w3c/web-platform-tests/css/css-values/vh-interpolate-pct.html imported/w3c/web-platform-tests/css/css-view-transitions/iframe-and-main-frame-transition-with-name-on-iframe.html imported/w3c/web-platform-tests/css/css-view-transitions/massive-element-on-top-of-viewport-offscreen-old.html (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/9518 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/70509 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/71585 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/15714 "Passed tests") | [❌ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/14628 "Found 2 new test failures: imported/w3c/web-platform-tests/webcodecs/temporal-svc-encoding.https.any.html?av1 imported/w3c/web-platform-tests/webcodecs/temporal-svc-encoding.https.any.worker.html?av1 (failure)") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/746 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12756 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/9286 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/14824 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/9124 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/12653 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/10932 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -397,6 +402,10 @@ - (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(N
+
+        [self willChangeValueForKey:@"_isBlockedByScreenTime"];
+        _isBlockedByScreenTime = urlIsBlocked;
+        [self didChangeValueForKey:@"_isBlockedByScreenTime"];
```
This should probably be inside the `if (urlWasBlocked != urlIsBlocked)`?

pxlcoder replied:
(this comment still hasn't been addressed)

jesxilin replied:
So for this, I noticed the`IsBlockedByScreenTimeKVOFalse` would fail if this code block was inside the `if`. This is b/c we start out with url _not_ blocked, then set it to url _not_ blocked again, so `_isBlockedByScreenTime` never changes and says it never updates. I was going to ask if I should 1) keep the current or 2) make the change and remove that test (I could also set urlIsBlocked to T then F but it would be basically identical to `IsBlockedByScreenTimeKVOTrue`.

pxlcoder replied:
> This is b/c we start out with url not blocked, then set it to url not blocked again, so _isBlockedByScreenTime never changes and says it never updates

In this case, it is expected that no updates are fired. So `IsBlockedByScreenTimeKVOFalse` is an incorrect test.

> (I could also set urlIsBlocked to T then F but it would be basically identical to IsBlockedByScreenTimeKVOTrue)

This test is still valuable as you'd be testing the T -> F transition, which is currently untested.

jesxilin replied:
Ok sounds good! Will change!

pxlcoder replied:
(you could just update `IsBlockedByScreenTimeKVOTrue` to test the T -> F transition right after the F -> T transition, and rename the test to `IsBlockedByScreenTimeKVO`)

pxlcoder reviewed (commented):
Please remove the empty file: `Tools/TestWebKitAPI/Tests/WebKitCocoa/Untitled.h`.

pxlcoder commented on Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
```diff
@@ -627,6 +627,8 @@ typedef NS_OPTIONS(NSUInteger, _WKWebViewDataType) {
 @property (nonatomic, copy, setter=_setEphemeralUIEventAttribution:) UIEventAttribution *_ephemeralUIEventAttribution WK_API_AVAILABLE(ios(16.0));
 - (void)_setEphemeralUIEventAttribution:(UIEventAttribution *)attribution forApplicationWithBundleID:(NSString *)bundleID WK_API_AVAILABLE(ios(16.0));
 
+@property (nonatomic, readonly, getter=_isBlockedByScreenTime) BOOL _isBlockedByScreenTime WK_API_AVAILABLE(macos(WK_MAC_TBA), ios(WK_IOS_TBA), visionos(WK_XROS_TBA));
```
`getter=_isBlockedByScreenTime` is unnecessary since this is `readonly` and the name already matches.

pxlcoder commented on Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
```diff
@@ -487,6 +488,7 @@ struct PerWebProcessState {
 - (RefPtr<WebKit::WebPageProxy>)_protectedPage;
 #if ENABLE(SCREEN_TIME)
 - (STWebpageController *)_screenTimeWebpageController;
+- (BOOL)_isBlockedByScreenTime;
```
This shouldn't be necessary, since it's already declared in the private header.

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ScreenTime.mm:
```diff
@@ -0,0 +1,88 @@
+@end
+
+@interface WKWebView (Internal)
+- (BOOL)_isBlockedByScreenTime;
```
Shouldn't be necessary since you have `<WebKit/WKWebViewPrivate.h>`.

pxlcoder commented on Tools/TestWebKitAPI/SourcesCocoa.txt:
```diff
@@ -378,3 +378,4 @@ Tests/WebKitCocoa/_WKInputDelegate.mm
 Tests/WebKitCocoa/_WKWebAuthenticationPanel.mm
 Tests/WebKitCocoa/iOSMouseSupport.mm
 Tests/WebKitCocoa/iOSStylusSupport.mm
+Tests/WebKitCocoa/ScreenTime.mm
```
Please put this in sorted order.

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ScreenTime.mm:
```diff
@@ -0,0 +1,88 @@
+ */
+#import "config.h"
+#if ENABLE(SCREEN_TIME)
+#import "TestWKWebView.h"
```
```suggestion

#import "config.h"

#if ENABLE(SCREEN_TIME)

#import "TestWKWebView.h"
```

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ScreenTime.mm:
```diff
@@ -0,0 +1,88 @@
+    [webView synchronouslyLoadHTMLString:@""];
+    EXPECT_FALSE([webView _isBlockedByScreenTime]);
+}
+#endif
```
```suggestion

#endif
```

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ScreenTime.mm:
```diff
@@ -0,0 +1,88 @@
+        if ([feature.key isEqualToString:@"ScreenTimeEnabled"])
+            [preferences _setEnabled:YES forFeature:feature];
+    }
+    RetainPtr controller = [webView _screenTimeWebpageController];
```
```suggestion

    RetainPtr controller = [webView _screenTimeWebpageController];
```

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ScreenTime.mm:
```diff
@@ -0,0 +1,88 @@
+
+TEST(ScreenTime, IsBlockedByScreenTimeMultiple)
+{
+    auto configuration = adoptNS([[WKWebViewConfiguration alloc] init]);
```
```suggestion
    RetainPtr configuration = adoptNS([[WKWebViewConfiguration alloc] init]);
```

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ScreenTime.mm:
```diff
@@ -0,0 +1,88 @@
+
+TEST(ScreenTime, IsBlockedByScreenTimeFalse)
+{
+    auto configuration = adoptNS([[WKWebViewConfiguration alloc] init]);
```
```suggestion
    RetainPtr configuration = adoptNS([[WKWebViewConfiguration alloc] init]);
```

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ScreenTime.mm:
```diff
@@ -0,0 +1,88 @@
+        if ([feature.key isEqualToString:@"ScreenTimeEnabled"])
+            [preferences _setEnabled:YES forFeature:feature];
+    }
+    RetainPtr controller = [webView _screenTimeWebpageController];
```
```suggestion

    RetainPtr controller = [webView _screenTimeWebpageController];
```

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ScreenTime.mm:
```diff
@@ -0,0 +1,88 @@
+
+TEST(ScreenTime, IsBlockedByScreenTimeTrue)
+{
+    auto configuration = adoptNS([[WKWebViewConfiguration alloc] init]);
```
```suggestion
    RetainPtr configuration = adoptNS([[WKWebViewConfiguration alloc] init]);
```

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ScreenTime.mm:
```diff
@@ -0,0 +1,88 @@
+        if ([feature.key isEqualToString:@"ScreenTimeEnabled"])
+            [preferences _setEnabled:YES forFeature:feature];
+    }
+    [webView synchronouslyLoadHTMLString:@""];
```
```suggestion

    [webView synchronouslyLoadHTMLString:@""];
    
```

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ScreenTime.mm:
```diff
@@ -0,0 +1,88 @@
+    [webView synchronouslyLoadHTMLString:@""];
+    RetainPtr controller = [webView _screenTimeWebpageController];
+    [controller setURLIsBlocked:YES];
+    EXPECT_TRUE([webView _isBlockedByScreenTime]);
```
```suggestion

    EXPECT_TRUE([webView _isBlockedByScreenTime]);
```

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ScreenTime.mm:
```diff
@@ -0,0 +1,88 @@
+    RetainPtr controller = [webView _screenTimeWebpageController];
+    [controller setURLIsBlocked:NO];
+    [webView synchronouslyLoadHTMLString:@""];
+    EXPECT_FALSE([webView _isBlockedByScreenTime]);
```
```suggestion

    EXPECT_FALSE([webView _isBlockedByScreenTime]);
```

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ScreenTime.mm:
```diff
@@ -0,0 +1,88 @@
+- (STWebpageController *)_screenTimeWebpageController;
+@end
+
+TEST(ScreenTime, IsBlockedByScreenTimeTrue)
```
I think we also need a test that tests KVO on the WKWebView property.

aprotyas commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ScreenTime.mm:
```diff
@@ -0,0 +1,88 @@
+    for (_WKFeature *feature in [WKPreferences _features]) {
+        if ([feature.key isEqualToString:@"ScreenTimeEnabled"])
+            [preferences _setEnabled:YES forFeature:feature];
+    }
```
Nit: Write a helper method like this?

```
static RetainPtr<TestWKWebView> webViewForScreenTimeTests()
{
    RetainPtr configuration = adoptNS([[WKWebViewConfiguration alloc] init]);
    auto preferences = [configuration preferences];
    for (_WKFeature *feature in [WKPreferences _features]) {
        if ([feature.key isEqualToString:@"ScreenTimeEnabled"])
            [preferences _setEnabled:YES forFeature:feature];
    }
    RetainPtr webView = adoptNS([[TestWKWebView alloc] initWithFrame:CGRectMake(0, 0, 400, 300) configuration:configuration.get()]);
    return webView;
}
```

aprotyas commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -386,6 +387,11 @@ - (void)_uninstallScreenTimeWebpageController
+- (BOOL)_isBlockedByScreenTime
+{
+    return _isBlockedByScreenTime;
+}
```
This needs to go under `@implementation WKWebView (WKPrivate)`

pxlcoder replied:
Also, since the instance variable is conditional on `ENABLE(SCREEN_TIME)`, but the property declaration is not, you probably want to do:

```
(BOOL)_isBlockedByScreenTime
{
#if ENABLE(SCREEN_TIME)
    return _isBlockedByScreenTime;
#else
    return NO;
#endif
}
```

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ScreenTime.mm:
```diff
@@ -0,0 +1,156 @@
+#import <wtf/RetainPtr.h>
+
+static void *blockedStateObserverChangeKVOContext = &blockedStateObserverChangeKVOContext;
+static bool stateDidChange;
```
```suggestion
static bool stateDidChange = false;
```

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ScreenTime.mm:
```diff
@@ -0,0 +1,156 @@
+
+@interface BlockedStateObserver : NSObject
+- (instancetype)initWithWebView:(TestWKWebView *)webView;
+@property (nonatomic) BOOL _isBlockedByScreenTime;
```
This property is unused, I think you can just remove it?

pxlcoder commented on Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
```diff
@@ -270,6 +270,7 @@ struct PerWebProcessState {
 #if ENABLE(SCREEN_TIME)
     RetainPtr<STWebpageController> _screenTimeWebpageController;
 #endif
+    BOOL _isBlockedByScreenTime;
```
This can be inside the `ENABLE`?

jesxilin replied:
Will change!

jesxilin committed:
[Screen Time Refactoring] Add WKWebView property to detect Screen Time blocking
https://bugs.webkit.org/show_bug.cgi?id=284703
rdar://140438960

Reviewed by Aditya Keerthi.

Created new API and property `_isBlockedByScreenTime` to indicate
if ScreenTime blocking is active.
Added API tests for new property.
Added missing header files for tests `SystemPreview.mm`.

* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _installScreenTimeWebpageController]):
(-[WKWebView observeValueForKeyPath:ofObject:change:context:]):
(-[WKWebView _isBlockedByScreenTime]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
* Tools/TestWebKitAPI/SourcesCocoa.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ScreenTime.mm: Added.
(webViewForScreenTimeTests):
(-[BlockedStateObserver initWithWebView:]):
(-[BlockedStateObserver observeValueForKeyPath:ofObject:change:context:]):
(TEST(ScreenTime, IsBlockedByScreenTimeTrue)):
(TEST(ScreenTime, IsBlockedByScreenTimeFalse)):
(TEST(ScreenTime, IsBlockedByScreenTimeMultiple)):
(TEST(ScreenTime, IsBlockedByScreenTimeKVO)):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SystemPreview.mm:

Canonical link: https://commits.webkit.org/288240@main
